### ISSUE TITLE
Ghost Websocket Connections!

### ISSUE THREAD

**Issue by dev-badace** [liveblocks/liveblocks#1459]
Dear @nvie , first of all I'd like to start with appreciating the level & amount of effort you've put in in implementing the websocket connection system based on a FSM, it truly is very impressive. I was going through your implementation today & may have found a bug/race condition (3 race conditions to trigger the same thing). I am not sure if it not an issue & left out intended since mostly the connection is handled by the library internally (leaving lesser scope for error). 

**Describe the bug**

They way the FSM currently handles the websocket connection & reconnection (even closes), does not accounts for the case where the connection may be accepted/open after the timeout is over, or the machine moves to a different state while the connection is being open. to be more precise I am talking about the `@connecting.busy` state of the machine.  

assume the given scenario: 
  - machine is in `@connecting.busy` state
  - a manual reconnect occurs at the same time
  - machine will now mark the current `@connecting.busy` as stale, and ignore it
  - machine moves to recreating a new connection, due to the manual reconnect
  - uh oh.. but the previous websocket connection attempt is still there leading to an zombie open socket

this above scenario can also happen with if the connection attempt resolves after the `SOCKET_CONNECT_TIMEOUT`. well this scenario is extremely unlikely since the variable is set to 10sec. I'd assume just like reconnect , a manual close can also cause this, haven't tried/tested that case yet but is extremely likely  

**To Reproduce**

there aren't very clear simple steps to reproduce this, but a few that I've tried to make it easier to see in action.

1. (easiest reproducibility)  Manually changing the `SOCKET_CONNECT_TIMEOUT` for the first time to a very low amount like 50, or any amount that is lesser than what it might actually take for the connection promise to resolve. example piece of code I added to lib code itself 
```typescript
      var SLOW_FIRST_TRY = () => {
        if(firstTry) {
          firstTry = false
          return 100
        }

        return SOCKET_CONNECT_TIMEOUT
      } 


      return withTimeout(
        connect$,
        SLOW_FIRST_TRY(),
        "Timed out during websocket connection"
      ).then(
 ```

2. (unpredictable reproducibility)  having a manual reconnect button which `onClick` calls `room.reconnect()`, this has some unpredictable reproducibility but should work. as a matter of fact I was able to generate >=20  ghost connections doing so. effectively reaching the room limit for my free plan. 
![dqwdqwdqwd](https://github.com/liveblocks/liveblocks/assets/58091764/33fa721c-6b1b-4ee7-a6d0-5f6b3fd79175)

**Expected behavior**

 Well my examples above may be devoid of actual real world usage of liveblocks, where this case may not even be there, there still exists a possibility of a this bug/race condition happening out there in the wild wild west. so Ideally these ghost connections should be taken care of and cleared/disconnected. 
 
 
 thankx for reading all this if you've made this far, have a great day, and a fantastic year. you people are doing really cool work ^^

**Comment by nvie** [comment]
Thanks for the kind words. Believe it or not, but I've run into this very issue two days ago myself, and I have been meaning to investigate it more deeply. I'll try to take a look at it tomorrow, because you're right, and this replication is amazingly helpful! 💯

**Comment by dev-badace** [comment]
I'm glad I was helpful. 😄 

this is an interesting one. I'd like to also share a few solutions that came to my mind,

1. taking a third cleanup function for stale success cases however this still can lead to `>1` open websocket connections till the window it's cleared by the cleanup function

2. a different variation of the above solution where the success case gets called with another argument boolean `isStale`, and checking for that, (hmm... i don't like this one).

the ideal place to cleanup would be when machine exists the `@connecting.busy`. so right before the state is marked stale 

```typescript
      return () => {
        cancelled = true;
        cleanup(this.current.context) //something like this, here we close the socket in context `tempSocket`
      };
 ```
maybe additional piece of info in the context like `tempSocket` that gets passed down to the cleanup via context.

**Comment by nvie** [comment]
This fix is now published with [v1.9.8](https://github.com/liveblocks/liveblocks/releases/tag/v1.9.8).
